File: lib/services/file_utils.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';
import '../models/anime.dart';

// For web platform
import 'package:universal_html/html.dart' if (dart.library.io) 'dart:io' as universal;

class FileUtils {
  static Future<Directory> getSaveDirectory() async {
    final appDir = await getApplicationDocumentsDirectory();
    final saveDir = Directory('${appDir.path}/saved_lists');
    
    if (!await saveDir.exists()) {
      await saveDir.create(recursive: true);
    }
    
    return saveDir;
  }
  
  static String formatFilename(String season, int year) {
    final now = DateTime.now();
    final timestamp = '${now.year}${now.month.toString().padLeft(2, '0')}${now.day.toString().padLeft(2, '0')}_'
                     '${now.hour.toString().padLeft(2, '0')}${now.minute.toString().padLeft(2, '0')}${now.second.toString().padLeft(2, '0')}';
    return 'anime_list_${season.toLowerCase()}_${year}_$timestamp.json';
  }
  
  static Future<String?> saveAnimeListWithSeason(List<Anime> animeList, String season, int year) async {
    try {
      final saveDir = await getSaveDirectory();
      final filename = formatFilename(season, year);
      final file = File('${saveDir.path}/$filename');
      
      // Convert anime list to JSON
      final jsonList = animeList.map((anime) => anime.toJson()).toList();
      
      // Write to file
      await file.writeAsString(jsonEncode(jsonList), flush: true);
      
      return filename;
    } catch (e) {
      print('Error saving anime list: $e');
      return null;
    }
  }
  
  static Future<List<Anime>> loadAnimeList(String filename) async {
    try {
      final saveDir = await getSaveDirectory();
      final file = File('${saveDir.path}/$filename');
      
      if (!await file.exists()) {
        return [];
      }
      
      final jsonString = await file.readAsString();
      final jsonList = jsonDecode(jsonString) as List;
      
      return jsonList.map((json) => Anime.fromJson(json)).toList();
    } catch (e) {
      print('Error loading anime list: $e');
      return [];
    }
  }
  
  static Future<List<String>> getSavedSeasonFiles() async {
    try {
      final saveDir = await getSaveDirectory();
      
      if (!await saveDir.exists()) {
        return [];
      }
      
      final files = await saveDir
          .list()
          .where((entity) => entity.path.endsWith('.json'))
          .map((entity) => entity.path.split('/').last)
          .toList();
      
      // Sort by most recent first
      files.sort((a, b) => b.compareTo(a));
      
      return files;
    } catch (e) {
      print('Error getting saved files: $e');
      return [];
    }
  }
  
  static List<Anime> sortAnimeList(List<Anime> animeList, String sortBy) {
    switch (sortBy) {
      case 'date':
        return List.from(animeList)..sort((a, b) => a.date.compareTo(b.date));
      case 'date_reverse':
        return List.from(animeList)..sort((a, b) => b.date.compareTo(a.date));
      case 'alpha':
        return List.from(animeList)..sort((a, b) => a.title.toLowerCase().compareTo(b.title.toLowerCase()));
      case 'alpha_reverse':
        return List.from(animeList)..sort((a, b) => b.title.toLowerCase().compareTo(a.title.toLowerCase()));
      case 'members_high':
        return List.from(animeList)..sort((a, b) => b.members.compareTo(a.members));
      case 'members_low':
        return List.from(animeList)..sort((a, b) => a.members.compareTo(b.members));
      default:
        return animeList;
    }
  }

  static Future<List<Anime>> readAnimeList() async {
    try {
      if (kIsWeb) {
        // Web implementation
        final storage = universal.window.localStorage;
        final jsonString = storage['anime_list'];
        
        if (jsonString == null || jsonString.isEmpty) {
          return [];
        }
        
        final List<dynamic> jsonList = json.decode(jsonString);
        return jsonList.map((json) => Anime.fromJson(json)).toList();
      } else {
        // Native implementation
        final directory = await getApplicationDocumentsDirectory();
        final file = File('${directory.path}/anime_list.json');
        
        if (!await file.exists()) {
          return [];
        }
        
        final contents = await file.readAsString();
        final List<dynamic> jsonList = json.decode(contents);
        
        return jsonList.map((json) => Anime.fromJson(json)).toList();
      }
    } catch (e) {
      print('Error reading anime list: $e');
      return [];
    }
  }
  
  static Future<void> saveAnimeList(List<Anime> animeList) async {
    try {
      final jsonList = animeList.map((anime) => anime.toJson()).toList();
      final jsonString = json.encode(jsonList);
      
      if (kIsWeb) {
        // Web implementation
        final storage = universal.window.localStorage;
        storage['anime_list'] = jsonString;
      } else {
        // Native implementation
        final directory = await getApplicationDocumentsDirectory();
        final file = File('${directory.path}/anime_list.json');
        await file.writeAsString(jsonString);
      }
    } catch (e) {
      print('Error saving anime list: $e');
    }
  }
  
  static Future<List<String>> getSavedFiles() async {
    if (kIsWeb) {
      // For web, just return the main anime list if it exists
      return universal.window.localStorage.containsKey('anime_list') 
          ? ['anime_list.json'] 
          : [];
    } else {
      final path = await getApplicationDocumentsDirectory();
      final dir = Directory(path.path);
      
      final List<String> files = [];
      await for (final entity in dir.list()) {
        if (entity is File && entity.path.endsWith('.json')) {
          files.add(entity.path.split('/').last);
        }
      }
      
      return files;
    }
  }
} 

File: python/file_utils.py
# File: utils/file_utils.py

import streamlit as st
import os
import json
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

def get_save_directory():
    """Get or create the directory for saved anime lists."""
    save_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'saved_lists')
    os.makedirs(save_dir, exist_ok=True)
    return save_dir

def format_filename(season, year):
    """Format filename with season, year, and timestamp."""
    current_time = datetime.now()
    timestamp = current_time.strftime("%Y%m%d_%H%M%S")
    return f"anime_list_{season.lower()}_{year}_{timestamp}.json"

def save_anime_list(anime_list, season, year):
    """Save anime list to a JSON file with formatted filename."""
    try:
        save_dir = get_save_directory()
        filename = format_filename(season, year)
        filepath = os.path.join(save_dir, filename)
        
        # Get updated state before saving
        from utils.data_utils import get_updated_anime_list
        
        # Ensure anime_list is not None and not empty
        if not anime_list:
            logger.error("No anime list to save")
            raise ValueError("No anime list to save")
            
        updated_list = get_updated_anime_list(anime_list)
        
        # Validate the updated list before saving
        if not updated_list:
            logger.error("Updated list is empty")
            raise ValueError("Updated list is empty")
            
        # Log some debugging info
        logger.info(f"Saving {len(updated_list)} anime entries")
        logger.info(f"First anime title: {updated_list[0].get('title', 'NO TITLE')}")
        
        # Ensure all required fields are present
        for anime in updated_list:
            if not anime.get('title'):
                logger.error(f"Missing title in anime entry: {anime}")
                continue
                
            # Ensure required fields with defaults
            if 'rssUrl' not in anime:
                anime['rssUrl'] = ''
            if 'fansubber' not in anime:
                anime['fansubber'] = 'ember'
        
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(updated_list, f, indent=2, ensure_ascii=False)
        
        logger.info(f"Successfully saved anime list to {filename}")
        return filename
    
    except Exception as e:
        logger.error(f"Error saving anime list: {str(e)}")
        # Log additional debug information
        logger.error(f"Session state keys: {st.session_state.keys()}")
        if 'edited_values' in st.session_state:
            logger.error(f"Edited values keys: {st.session_state.edited_values.keys()}")
        raise

def load_anime_list(filename):
    """Load anime list from a JSON file."""
    try:
        filepath = os.path.join(get_save_directory(), filename)
        with open(filepath, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        logger.error(f"Error loading anime list: {str(e)}")
        raise

def get_saved_files():
    """Get list of saved anime list files."""
    try:
        save_dir = get_save_directory()
        files = [f for f in os.listdir(save_dir) if f.endswith('.json')]
        return sorted(files, reverse=True)  # Most recent first
    except Exception as e:
        logger.error(f"Error getting saved files: {str(e)}")
        return []

def sort_anime_list(anime_list, sort_by):
    """Sort anime list based on specified criteria."""
    if sort_by == "date":
        return sorted(anime_list, key=lambda x: x.get('date', ''))
    elif sort_by == "date_reverse":
        return sorted(anime_list, key=lambda x: x.get('date', ''), reverse=True)
    elif sort_by == "alpha":
        return sorted(anime_list, key=lambda x: x.get('title', '').lower())
    elif sort_by == "alpha_reverse":
        return sorted(anime_list, key=lambda x: x.get('title', '').lower(), reverse=True)
    elif sort_by == "members_high":
        return sorted(anime_list, key=lambda x: x.get('members', 0), reverse=True)
    elif sort_by == "members_low":
        return sorted(anime_list, key=lambda x: x.get('members', 0))
    return anime_list